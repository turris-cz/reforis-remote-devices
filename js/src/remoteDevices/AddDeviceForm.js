/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    useAPIPost,
    API_STATE,
    useForm,
    useAlert,
    FileInput,
    Button,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export default function AddDeviceForm() {
    const [setAlert] = useAlert();

    const [postDeviceResponse, postDevice] = useAPIPost(
        API_URLs.devices,
        "multipart/form-data"
    );
    useEffect(() => {
        if (postDeviceResponse.state === API_STATE.ERROR) {
            setAlert(postDeviceResponse.data);
        }
    }, [postDeviceResponse, setAlert]);
    const isSending = postDeviceResponse.state === API_STATE.SENDING;

    const [formState, formChangeHandler, reloadForm] = useForm(validator);
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        reloadForm({ token: undefined });
    }, [reloadForm]);

    function handleSubmit(event) {
        event.preventDefault();
        const postData = new FormData();
        postData.append("token", formData.token);
        postDevice({ data: postData });
    }

    if (!formData) {
        return null;
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Add Device")}</h2>
            <p>
                {_(
                    "Upload token file generated by device you want to control remotely."
                )}
            </p>
            <form onSubmit={handleSubmit} className="position-relative">
                <FileInput
                    label={
                        formData.token ? formData.token.name : _("Token file")
                    }
                    files={[formData.token]}
                    error={formErrors.token}
                    onChange={formChangeHandler((value) => ({
                        token: { $set: value },
                    }))}
                    accept=".tar.gz,.tgz"
                />
                <div className="text-right">
                    <Button
                        type="submit"
                        forisFormSize
                        disabled={
                            formErrors.token || !formData.token || isSending
                        }
                        loading={isSending}
                    >
                        {_("Upload token")}
                    </Button>
                </div>
            </form>
        </div>
    );
}

function validator(formData) {
    // Ignore empty file - submit will be disabled anyways
    if (!formData.token) {
        return undefined;
    }

    if (formData.token.size > 1024 * 1024) {
        return { token: _("File is too big. Maximum size is 1 MB.") };
    }

    const filename = formData.token.name;
    if (filename.length === 0 || filename.length > 50) {
        return {
            token: _(
                "Filename must be at least 1 and at most 50 characters long."
            ),
        };
    }
    if (!/^[a-zA-Z0-9_.-]*$/.test(filename)) {
        return {
            token: _(
                "Filename can contain only alphanumeric characters, dots, dashes and underscores."
            ),
        };
    }

    return undefined;
}
