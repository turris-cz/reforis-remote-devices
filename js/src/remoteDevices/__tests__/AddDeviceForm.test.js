/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import {
    render,
    fireEvent,
    getByLabelText,
    getByText,
    wait,
} from "foris/testUtils/customTestRender";

import AddDeviceForm from "../AddDeviceForm";

function getFormElements(container) {
    return {
        fileInput: getByLabelText(container, "Choose token file"),
        submitButton: getByText(container, "Upload token"),
    };
}

function setFilesProperty(input, files) {
    // Files can't be set by fireEvent, see https://github.com/testing-library/react-testing-library/issues/93
    Object.defineProperty(input, "files", { value: files, configurable: true });
}

describe("<AddDeviceForm />", () => {
    let container;
    let fileInput;
    let submitButton;
    const emptyFile = new File([], "token.tar.gz");

    beforeEach(() => {
        ({ container } = render(<AddDeviceForm />));
        ({ fileInput, submitButton } = getFormElements(container));
    });

    it("should render form", () => {
        expect(container).toMatchSnapshot();
    });

    it("should add device", () => {
        expect(submitButton.disabled).toBe(true);
        setFilesProperty(fileInput, [emptyFile]);
        fireEvent.change(fileInput);
        expect(submitButton.disabled).toBe(false);

        fireEvent.click(submitButton);
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/remote-devices/api/devices",
            expect.anything(), // Can't compare two FormData objects
            expect.anything()
        );
    });

    it("should validate form", () => {
        // Wrong length of file name
        setFilesProperty(fileInput, [new File([], "q".repeat(51))]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                container,
                "The filename must be at least 1 and at most 50 characters long."
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // Invalid characters
        setFilesProperty(fileInput, [new File([], "!@#!$@%.tar.gz")]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                container,
                "The filename can contain only alphanumeric characters, dots, dashes, and underscores."
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // File too big
        setFilesProperty(fileInput, [
            new File(["q".repeat(1024 * 1025)], "turris.tar.gz"),
        ]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                container,
                "The file is too big. The maximum size is 1 MB."
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should handle API error", async () => {
        setFilesProperty(fileInput, [emptyFile]);
        fireEvent.change(fileInput);
        fireEvent.click(submitButton);

        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });
});
