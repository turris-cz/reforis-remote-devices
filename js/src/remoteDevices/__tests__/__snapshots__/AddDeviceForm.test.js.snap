// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddDeviceForm /> should render form 1`] = `
<div>
  <form>
    <h3>
      Add Device
    </h3>
    <p>
      Upload token file generated by device you want to control remotely.
    </p>
    <div
      class="form-group col-sm-12 offset-lg-1 col-lg-10"
    >
      <label
        class="custom-file-label"
        for="1"
      >
        Token file
      </label>
      <div
        class="input-group custom-file"
      >
        <input
          accept=".tar.gz,.tgz"
          class="form-control custom-file-input"
          files=""
          id="1"
          type="file"
        />
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      disabled=""
      type="submit"
    >
       
       
      Upload token
    </button>
  </form>
</div>
`;
